### ISSUE TITLE
Parallel Tests Not Thread-Safe

### ISSUE THREAD

**Issue by BlythMeister** [nunit/nunit#2573]
Given this as a (stupid) example
```
[TestFixture, Parallelizable(ParallelScope.All)]
    public class MockupTests2
    {
        private int counter;
        [SetUp]
        public void Setup()
        {
            counter = 0;
        }
        [Test]
        public void Test()
        {
            counter++;
            Thread.Sleep(10);
            Assert.AreEqual(1, counter);
        }
        [Test]
        public void Test2()
        {
            counter++;
            Thread.Sleep(10);
            Assert.AreEqual(1, counter);
        }
    }
```
These tests will fail every time.

The behavior i expect from NUnit is that each parallel execution of a test has it's own instance of the test class to ensure thread safety of non static fields.

It would appear (not looked at code, just reverse engineered) that NUnit has a single instance of the class and calls the setup and test on different threads.

**Comment by jnm2** [comment]
This would be a breaking change in behavior, so we'd need a way to opt in. See the existing discussion at https://github.com/nunit/nunit/issues/2105.

@nunit/framework-team People keep asking for this and more than one duplicate issue have even cropped up. I think we should listen.

**Comment by CharliePoole** [comment]
Yes, NUnit has always had a single instance per fixture, used for all the tests in that fixture. It was - back in 2000 - a substantial departure from JUnit practice, but seemed right to the developers at the time. In the past (pre-parallelism) the same possibility of corruption between tests existed, for example, if the initialization occured in the constructor. We would simply tell people "Don't do that!" but rather use a SetUp method. The general guidance was to keep code out of the constructor and rely on SetUp.

When a test is run in parallel, your setup naturally repeats itself multiple times. You really have changeable state in the object. Were we redesigning from scratch, a separate instance per test would be the sensible approach. But of course, we aren't designing from scratch. ☹️

We can still answer you with "Don't do that!" and it will work for the simple example. However, it may become inconvenient for many real examples. Currently, we would have to tell you "Don't allow such tests (with modifiable state) to run in parallel." That's the state of the software right now.

I agree that it would be convenient if we automatically ran test cases in separate instances in certain situations. Figuring out those situations without breaking somebody else's tests is the hard part. Some folks have already adapted to the fact that we run tests in the same instance and have code that depends on it. Additionally, I don't like the idea that we would change the life-cycle of the object simply because we were running the tests in parallel. You might make a change at the assembly level and then see your tests all start to work in a different way.

All this is just to say that it's a harder problem than you may have considered. We will probably want to introduce it as a new, opt-in feature rather than simply changing existing behavior.

@jnm2 Thanks for reminding me about #2105. Seems like I pretty much said the same thing on that as on this one. Short story: I think we should do something, but not the specific something asked for (automatic change of life cycle) on both issues. You and Chris seemed to agree. However, I see the issue was closed without action and I can't find any other issue that says what we may want to do about this. 

As a "lame duck" I have been trying to sit out major decisions of direction by merely commenting rather than taking any action, but it's hard. 😕 I agree with you that __something__ should be done about this. My thinking is that allowing changes to the life-cycle on a fixture-by-fixture basis complicates things too much, both for us and the user. I'd like to see a top-level setting or an assembly attribute that indicates which life-cycle to use. I lean more toward an assembly attribute, since you would generally __always__ want to use the best setting for the tests in a given assembly.

**Comment by BlythMeister** [comment]
seems this is a duplicate of #2105 
And the solution is to not use Setup and tear down in parallel tests.
Also, don't put test state outside of the test.

Which actually makes sense as it's the same as any code being executed on threads.
Fields are never thread safe (without funky attributes)

**Comment by CharliePoole** [comment]
Since #2105 is closed without action, I think one of us on the @nunit/framework-team needs to create an issue that reflects what we are __actually__ willing to do to resolve the problem - or retitle and reopen this one.

I'm proposing
 * An assembly level attribute that allows setting the way test fixtures are created to be single instance or instance per fixture.
 * Single Instance would work just like now
 * Instance per fixture would create a new fixture for each test
 * EITHER Perform one time setup and teardown activities for each test (i.e. no longer "one time") OR cause OneTimeSetUp and OneTimeTearDown to be flagged as errors.

Possible extensions:
 * Make this class level as well, allowing individual fixtures to have different life cycles.
 * Have a third assembly-level setting that automatically runs separate instances for any fixture containing one or more tests that are run in parallel. (Tricky to always know in advance)

I'd say do the basics first, however.
 *

**Comment by jnm2** [comment]
Thanks @BlythMeister for bringing this to our attention. It prompted me to file https://github.com/nunit/docs/issues/242.

I'm also going to start a new issue on the design of an instance-per-test-case feature since #2105 is quite long.

**Comment by jnm2** [comment]
@CharliePoole Thanks for that. I commented at the same instant somehow! 😆 I brought your proposal to https://github.com/nunit/nunit/issues/2574.

**Comment by janissimsons** [comment]
Why was this issue closed? Have you implemented it?

**Comment by BlythMeister** [comment]
Because it's a duplicate as i put in the comment before I closed.

Also, it's working as currently designed.

The way I see it, you have to make your test class thread safe (as you do with ANY other class)
If you new up a class with 2 methods, and a field being updated by both.
You are going to see the same behaviour as nunit.

As developers (me included as I raised this) need to think more about how tests are written to ensure that the test fixtures themselves are thread safe.
Each fixture class is essentially a singleton.

The best way to make your tests thread safe, do not use any fields in the fixture, but have a "Test context" class with that state that you create at the start of each test (a kind of "SetUp")
